File: ios/Plugin/EmailComposerPlugin.m
#import <Foundation/Foundation.h>
#import <Capacitor/Capacitor.h>

// Define the plugin using the CAP_PLUGIN Macro, and
// each method the plugin supports using the CAP_PLUGIN_METHOD macro.
CAP_PLUGIN(EmailComposerPlugin, "EmailComposer",
           CAP_PLUGIN_METHOD(hasAccount, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(open, CAPPluginReturnPromise);
)

File: ios/Plugin/EmailComposerPlugin.swift
import Foundation
import Capacitor
import MessageUI

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(EmailComposerPlugin)
public class EmailComposerPlugin: CAPPlugin, MFMailComposeViewControllerDelegate {
    private let implementation = EmailComposer()

    private var savedOpenCall: CAPPluginCall?

    @objc func hasAccount(_ call: CAPPluginCall) {
        call.resolve(["hasAccount": self.implementation.canSendMail()])
    }

    @objc func open(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            do {
                self.savedOpenCall = call
                let draft = try self.implementation.getMailComposerFromCall(call, delegateTo: self)
                self.bridge?.viewController?.present(draft, animated: true, completion: nil)
            } catch {
                call.reject(error as! String)
            }
        }
    }

    public func mailComposeController(_ controller: MFMailComposeViewController, didFinishWith result: MFMailComposeResult, error: Error?) {
        controller.dismiss(animated: true, completion: nil)
        if self.savedOpenCall != nil {
            self.savedOpenCall?.resolve()
        }
    }
}
